pkg/kubelet/dockertools/container_gc.go

```diff
@@ -245,35 +281,56 @@ func (cgc *containerGC) GarbageCollect(gcPolicy kubecontainer.ContainerGCPolicy,
 	return nil
 }
 
-func (cgc *containerGC) removeContainer(id string, podNameWithNamespace string, containerName string) {
-	glog.V(4).Infof("Removing container %q name %q", id, containerName)
-	err := cgc.client.RemoveContainer(id, dockertypes.ContainerRemoveOptions{RemoveVolumes: true})
+func (cgc *containerGC) netContainerCleanup(containerInfo containerGCInfo) {
```

Comment:
@caseydavenport @freehan I was going to do a follow-on to this for the docker runtime that ensured that a previous infra container *was* cleaned up before starting a new one.  Currently the docker runtime only cleans up running infra containers:

```
func (dm *DockerManager) SyncPod(...) {
	<snip>
	if containerChanges.StartInfraContainer || (len(containerChanges.ContainersToKeep) == 0 && len(containerChanges.ContainersToStart) == 0) {
		if len(containerChanges.ContainersToKeep) == 0 && len(containerChanges.ContainersToStart) == 0 {
			glog.V(4).Infof("Killing Infra Container for %q because all other containers are dead.", format.Pod(pod))
		} else {
			glog.V(4).Infof("Killing Infra Container for %q, will start new one", format.Pod(pod))
		}

		killResult := dm.killPodWithSyncResult(pod, kubecontainer.ConvertPodStatusToRunningPod(dm.Type(), podStatus), nil)
```

unfortunately that kubecontainer.ConvertPodStatusToRunningPod() call only converts running containers, so the dead infra is left to GC and not cleaned up.  Perhaps if we also solve this, it will fix the possible conflict with plugin workarounds by ensuring that the dead infra always gets cleaned up before a new one is started, as you suggest?